File: Top_Interview_Questions/2_AddTowNumbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* ans = new ListNode(0);
        ListNode* dummy = ans;
        int carry = 0;
        
        while(l1 || l2) {
            //需要检查是否存在，不能直接访问
            int val1 = l1 ? l1->val : 0;
            int val2 = l2 ? l2->val : 0;
            int temp = val1 + val2 + carry;
            carry = temp / 10;
            ans->val = temp % 10;
            
            if(l1) l1 = l1->next;
            if(l2) l2 = l2->next;
            
            if(l1 || l2) {
                ans->next = new ListNode(0);
                ans = ans->next;
            }
        }
        
        if(carry == 1)
            ans->next = new ListNode(1);
        return dummy;
    }
};

File: Train/go/2_AddTowNumbers/2_AddTowNumbers.go
package AddTowNumbers

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var dummy ListNode
	cur := &dummy
	carry := 0

	for l1 != nil || l2 != nil || carry != 0 {
		sum := carry
		if l1 != nil {
			sum += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			sum += l2.Val
			l2 = l2.Next
		}
		carry = sum / 10
		sum %= 10
		cur.Next = &ListNode{
			Val: sum,
		}
		cur = cur.Next
	}
	return dummy.Next
}
